### ISSUE TITLE
Async version of functions never return

### ISSUE THREAD

**Issue by CorrM** [xamarin/xamarin-android#4272]
Async version of function [WaitFor](https://docs.microsoft.com/en-us/dotnet/api/java.lang.process.waitforasync?view=xamarin-android-sdk-9) and [ReadToEndAsync](https://docs.microsoft.com/en-us/dotnet/api/system.io.streamreader.readtoendasync?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev16.query%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(System.IO.StreamReader.ReadToEndAsync);k(TargetFrameworkMoniker-MonoAndroid,Version%3Dv6.0);k(DevLang-csharp)%26rd%3Dtrue&view=netframework-4.8)

this function never return even the process has exit !!

Example :
```
public static async Task<string> Exec(string cmd)
{
	// Create su process
	Java.Lang.Process sh = Java.Lang.Runtime.GetRuntime().Exec("su -c " + "cat /data/misc/vpn/state");

	int exitCode = await sh.WaitForAsync(); // => Here never return

	string ret;
	if (sh.ErrorStream != null)
	{
		using var errorStreamReader = new StreamReader(sh.ErrorStream);
		ret = await errorStreamReader.ReadToEndAsync();  // => Here never return
	}
	else
	{
		using var outputStreamReader = new StreamReader(sh.InputStream);
		ret = await outputStreamReader.ReadToEndAsync();  // => Here never return
	}

	Log.WriteLine(LogPriority.Debug, "Su", ret); // Never hit
	return ret;
}
```

if i used `WaitFor` and `ReadToEnd` it's run normal.
Async version never return.

**Comment by grendello** [comment]
@CorrM when you run with the synchronous version, do you get a password prompt from `su`? This is the first thought that comes to mind, that the async version (running in a different thread) doesn't properly interact with `su` to provide it with the password it most probably needs and waits for. `Java.Lang.Process.WaitForAsync` is a simple wrapper around `WaitFor`:

```csharp
public global::System.Threading.Tasks.Task<int> WaitForAsync ()                                                                                                                                                                                                                                             
{
   return global::System.Threading.Tasks.Task.Run (() => WaitFor ());
}
```

Also, if `WaitForAsync` never returns, how do you get to the other places when functions never return? Is it when you use `WaitFor` instead of `WaitForAsync`?

Instead of using `WiatForAsync`, copy the above `Task.Run` invocation to your code and see if `WaitFor` inside the task actually returns.

Unfortunately, there's not enough information provided to diagnose what else might be going on. Please try the above suggestions and if you don't find an explanation, please provide:

 * logcat output of the failed session (full, run `adb logcat -c` before starting your app to clear the buffer)
 * a minimal application that reproduces the issue
 * information on where you run your code (emulator - which API level; device(s) - API levels, models etc)
 * Xamarin.Android version you use

**Comment by CorrM** [comment]
`do you get a password prompt.?` > No.
`Is it when you use WaitFor instead of WaitForAsync?` > Yes.
`Xamarin.Android` > 10
`emulator` > Memu, Android 7.1 - Api 25
`Instead of using WiatForAsync, copy the above Task.Run` > Works !!, and test again the original method it's works now.

as i think the problem was because the su exec method called from [ MainActivaty =>Button Click Event (async void) => Su Exec function ].
now i use it like that [ Main Activaty => Button Click Event => Task.Run => Su Exec function ] and it's work.

why that happen .?

**Comment by CorrM** [comment]
@grendello

**Comment by grendello** [comment]
I can't tell for sure, but I suppose it had to do with the fact that an `async void` method was involved. `async void` [should be avoided](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void) because of the differences it has with those async methods returning `Task` or `Task<T>`.

**Comment by CorrM** [comment]
@grendello anyway it's works for me, or i get arround that by create a new task and do what i want.
so, if u want to close do it.
